pgx/src/lib.rs

```diff
@@ -128,6 +128,20 @@ pub use pg_sys::{
 #[doc(hidden)]
 pub use pgx_sql_entity_graph;
 
+// Postgres v15 has the concept of an ABI "name".  The default is `b"PostgreSQL\0"` and this is the
+// ABI that pgx extensions expect to be running under.  We will refuse to compile if it is detected
+// that we're trying to be built against some other kind of "postgres" that has its own ABI name.
+//
+// Unless the compiling user explicitly told us that they're aware of this via `--features unsafe-postgres`.
+#[cfg(all(feature = "pg15", not(feature = "unsafe-postgres")))]
+const _: () = {
+    if crate::pg_sys::FMGR_ABI_EXTRA.iter().zip(b"PostgreSQL\0".iter()).all(|(a, b)| a == b) {
+        compile_error!(
+            "Unrecognized Postgres ABI.  Perhaps you need `--features unsafe-postgres`?"
```

Comment:
We should error if it doesn't match, not if it does match.
```suggestion
    if !crate::pg_sys::FMGR_ABI_EXTRA.iter().zip(b"PostgreSQL\0".iter()).all(|(a, b)| a == b) {
        compile_error!(
            "Unrecognized Postgres ABI.  Perhaps you need `--features unsafe-postgres`?"
```

Reply:
Yeah, I'm learning how to program today.  See latest push.